include/HAL/TargetSystem.h

```diff
@@ -156,6 +183,12 @@ class Target {
 
 private:
   mlir::TimingScope rootTimer;
+
+  /// @brief List of diagnostics generated for this target
+  qssc::DiagList diagnostics_;
+
+  /// @brief Mutex for adding diagnostics to the diagnostic list
+  std::mutex diagnosticsMutex_;
```

Comment:
This could probably be a shared_mutex with exclusive locking only for adding a diagnostic?

Reply:
On further reflection I don't think that shared_mutex makes sense in this context. Diagnostics may be added in a threaded manner, but they should only ever be read in a single thread because the callbacks that emit them may not be thread safe.